csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -474,7 +474,7 @@ class ScopeMap : private kir::IrVisitor {
   void handle(kir::ForLoop* for_loop) final {
     auto loop_info = makeAndRegisterScopeInfo(for_loop);
     kir::IrVisitor::handle(for_loop);
-    loop_info->end_pos = expr_pos_map_.getCurrentPos() + 1;
+    loop_info->end_pos = expr_pos_map_.getCurrentPos();
```

Comment:
Thanks for the fix. I wonder if we should instead do:

```suggestion
    expr_pos_map_.moveToNext();
    loop_info->end_pos = expr_pos_map_.getCurrentPos();
```

This would effectively assign its own "line" to each `EndFor`. It seems to me this would more faithfully represent the enclosed region of each loop.

Reply:
I like that approach since it's a bit strange to have an inner expression for one endpoint of an interval whose start endpoint is the for loop. However, I have assumed that when we traverse we will pass an expression corresponding to the first write and last read of each interval,which would no longer be the case with this change. That can be fixed but it's just a potential complication.

Reply:
> However, I have assumed that when we traverse we will pass an expression corresponding to the first write and last read of each interval,which would no longer be the case with this change. 

Could you give me some simple example?

Reply:
In `PromoteReuseSyncModifier`, we traverse expressions to look for syncs in the sync intervals. The last read determines the start of a sync interval. When we traverse to an Expr, we look up its position to see if it is the start of a sync interval, and if so we add that sync interval's right endpoint to a set that we can also check. The right endpoints will always correspond to Exprs, but with this change the left endpoints do not, so we would need to be careful that these don't get skipped, since that would effectively skip inserting the sync.

Reply:
In that case, we could rewrite that class to traverse to find pre-existing syncing exprs, then iterate over sync intervals in order of right endpoint, checking for existing syncs within that inverval or inserting them as needed. A method like that would only depend on the right endpoints of the sync intervals belonging to actual `Expr`s, which they would since they are always first write expressions.

Reply:
Is this:

> but with this change the left endpoints do not

because

>  The last read determines the start of a sync interval. 

?

Isn't the last read determines the end of the last point of the live interval, and shouldn't the start of its sync interval start at +1 position?

Reply:
You're right. In that case I can just increment the start interval inside that check. I'll try it on and see how it fits tomorrow morning.